### PR TITLE
Added symbol support

### LINKED ISSUE #20138: Remove `index_of` from `DiscreteMarkovChain`
After merging of https://github.com/sympy/sympy/pull/20042 named states such as `Symbol` objects are required to be accessed by `index_of` as shown by the example below,

https://github.com/sympy/sympy/blob/b88c69a1edf941c211a49fee9a8a56be790f4196/sympy/stats/stochastic_process_types.py#L688-L696

The above should be made as simple as the following,

```python
     >>> sunny, cloudy, rainy = symbols('Sunny, Cloudy, Rainy') 
     >>> Y = DiscreteMarkovChain("Y", [sunny, cloudy, rainy], T) 
     >>> P(Eq(Y[3], rainy), Eq(Y[1], cloudy)).round(2) 
     0.36
  
     >>> E(Y[3], Eq(Y[1], cloudy)) 
     0.38*Cloudy + 0.36*Rainy + 0.26*Sunny 
```

If possible try doing the same for strings as well but only after making the stuff possible.

### PR DESCRIPTION
**Pull request by naveensaigit**
Fix #20138 

#### Brief description of what is fixed or changed
I've added a function which removes index_of for symbols

#### Other comments
To be done for strings as well

#### Release Notes

* stats
    * The state space of `ContinuousMarkovChain` is now by default a `Range` object instead of `S.Reals`.
    * The generator matrix of `ContinuousMarkovChain` is now by default a `MatrixSymbol` object instead of `None`.

### CODE REVIEW

**Comment by naveensaigit** [comment]
@czgdp1807 Please check this out

**Comment by czgdp1807** [comment]
Can you modify the tests in the [test file](https://github.com/sympy/sympy/blob/master/sympy/stats/tests/test_stochastic_process.py) as well?

File: sympy/stats/stochastic_process_types.py [hunk 1]
```diff
@@ -421,6 +421,15 @@ def _preprocess(self, given_condition, evaluate):
 
         return is_insufficient, trans_probs, state_index, given_condition
 
+    def replace_with_index(self, condition):
+        if isinstance(condition, Relational):
+            if condition.rhs.is_symbol:
+                try:
+                    return Eq(condition.lhs, self.state_space.index(condition.rhs))
+                except NameError:
+                    raise NameError('%s not found in the state space'%condition.rhs)
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:430]:
```suggestion
        if isinstance(condition, Relational):
            return type(condition)(self.state_space.index(condition.lhs), self.state_space.index(condition.rhs))
```

Comment (czgdp1807) [reply sympy/stats/stochastic_process_types.py:430]: I also see that, the implementation of `index_of` in `master` is a bit inefficient,

https://github.com/sympy/sympy/blob/bdb49c4abfb35554a3c8ce761696ffff3bb837fe/sympy/stats/stochastic_process_types.py#L753-L766

The dictionary should be computed only once and stored as one of the attributes of the object. Can you please optimise it in your PR?

Comment (naveensaigit) [reply sympy/stats/stochastic_process_types.py:430]: Yeah, I'll fix that. I was wondering if `self.state_space.index_of[condition.lhs]` is required as lhs would remain the same.

Comment (naveensaigit) [reply sympy/stats/stochastic_process_types.py:430]: Lastly, for strings, right now, its possible to pass Eq(Y[3],Str('Cloudy')) and we can convert this to index in replace_with_index. Would that be okay?

Comment (czgdp1807) [reply sympy/stats/stochastic_process_types.py:430]: > I was wondering if `self.state_space.index_of[condition.lhs]` is required as lhs would remain the same.

What about, `Gt(2, Y[3])` or `Eq(Str('Cloudy'), Y[3])` or in fact, how does this work, `Eq(Y[5], Y[6])`?

Comment (naveensaigit) [reply sympy/stats/stochastic_process_types.py:430]: > What about, `Gt(2, Y[3])` or `Eq(Str('Cloudy'), Y[3])` or in fact, how does this work, `Eq(Y`

Oh, I'll add an `isinstance(Y[i],RandomIndexedSymbol)` check to make sure I replace the right side of the equation

**Comment by naveensaigit** [comment]
I have added tests and made the code suitable for expressions like `Gt(2, Y[3]), Le(Y[4], cloudy)`. However, it looks like Eq(Y[2], Y[3]) gives 
``` python
NotImplementedError : Sorry, as_set has not yet been implemented for multivariate expressions
```
I'll raise another issue for that. 
Also, using Eq(Str('Cloudy'), Y[1]) gives 
``` python
TypeError: Can only compare inequalities with Expr
```
Otherwise, I think all other cases work

**Comment by czgdp1807** [comment]
> However, it looks like Eq(Y[2], Y[3]) gives

Actually, this seems to be a deeper issue, though queries of these type may not be that common but supporting such things would be great and modifying/extending the current algorithm can be challenging. Consider the following example,

```python
In [1]: from sympy.stats import *

In [2]: T = Matrix([[S(9)/10, S(1)/10], [S(1)/2, S(1)/2]])

In [3]: M = DiscreteMarkovChain('M', [0, 1], T)

In [4]: P(Eq(M[2], M[3]), Eq(M[0], 1))
``` 
We are trying to find the probability of the state at time 2 and time 3 to be equal given at time 0 the state was 1. The first approach that comes to my mind is to consider all values of `M[2]` and `M[3]` for which `Eq(M[2], M[3])` holds which are clearly `(1, 1)` and `(0, 0)`. Now, answering, `P(Eq(M[2], M[3]), Eq(M[0], 1))` will be equivalent to answering, `P(Eq(M[2], 1) & Eq(M[3], 1), Eq(M[0], 1)) + P(Eq(M[2], 0) & Eq(M[3], 0), Eq(M[0], 1))`. Though, this is not generic enough for cases like `Gt(M[2], M[3])` where finding required tuples can be challenging and then passing transformed queries. Another way is to do `P(Eq(M[2], M[3]), Eq(M[0], 1) & Eq(M[2], si) & Eq(M[3], sj))` where si, sj are the values from state space. This doesn't work as of now but can be made to work by utilising the given conditions fully before applying `as_set` on the `Relational` objects in the query. The issue for every query of type `Relational(RandomIndexedSymbol, RandomIndexedSymbol)` we would be iterating `O(size_of_state_space^2)` values which lead to overall complexity `O(time_for_single_query*size_of_state_space^2*number_of_queries_of_above_type) + O(time_for_single_query)`. The first can be zero though if no such queries are present. Optimising the first part is the task that may need attention.

**Comment by naveensaigit** [comment]
> The issue for every query of type `Relational(RandomIndexedSymbol, RandomIndexedSymbol)` we would be iterating `O(size_of_state_space^2)` values which lead to overall complexity 

Yes, that's true. I'll check if this can be done in any other way, or we'll proceed with this approach.

**Comment by czgdp1807** [comment]
> Yes, that's true. I'll check if this can be done in any other way, or we'll proceed with this approach.

Yeah, but in a different PR. Though instead of passing `Eq(M[2], 1) & Eq(M[3], 1)` we can just substitute the Relational objects before making a recursive call, i.e., `Eq(M[2], M[3]).subs({M[2]: si, M[3]: sj})`. Whatever minor changes in implementation will be needed. Maybe for optimising this, one direction that can be take is to reduce the number of `Relational(RandomIndexedSymbol, RandomIndexedSymbol)` objects in the query before starting to iterate all the possible tuples in the state space. It will be a bit tricky. Keep thinking on this.

**Comment by naveensaigit** [comment]
> > Yes, that's true. I'll check if this can be done in any other way, or we'll proceed with this approach.
> 
> Yeah, but in a different PR. Though instead of passing `Eq(M[2], 1) & Eq(M[3], 1)` we can just substitute the Relational objects before making a recursive call, i.e., `Eq(M[2], M[3]).subs({M[2]: si, M[3]: sj})`. Whatever minor changes in implementation will be needed. Maybe for optimising this, one direction that can be take is to reduce the number of `Relational(RandomIndexedSymbol, RandomIndexedSymbol)` objects in the query before starting to iterate all the possible tuples in the state space. It will be a bit tricky. Keep thinking on this.

Yes, I'll open another issue for this and think of faster methods meanwhile

File: sympy/stats/stochastic_process_types.py [hunk 2]
```diff
@@ -421,6 +422,20 @@ def _preprocess(self, given_condition, evaluate):
...
+            if isinstance(condition.rhs, Symbol) or isinstance(condition.rhs, Str):
+                try:
+                    condition = type(condition)(condition.lhs, self.index_of[condition.rhs])
+                except NameError:
+                    raise NameError('%s not found in the state space'%condition.rhs)
+            if isinstance(condition.lhs, Symbol) or isinstance(condition.rhs, Str):
+                try:
+                    condition = type(condition)(self.index_of[condition.lhs], condition.rhs)
+                except NameError:
+                    raise NameError('%s not found in the state space'%condition.lhs)
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:436]:
```suggestion
                condition = type(condition)(self.index_of[condition.lhs], condition.rhs)
```

File: sympy/stats/stochastic_process_types.py [hunk 3]
```diff
@@ -421,6 +422,20 @@ def _preprocess(self, given_condition, evaluate):
 
         return is_insufficient, trans_probs, state_index, given_condition
 
+    def replace_with_index(self, condition):
+        if isinstance(condition, Relational):
+            if isinstance(condition.rhs, Symbol) or isinstance(condition.rhs, Str):
+                try:
+                    condition = type(condition)(condition.lhs, self.index_of[condition.rhs])
+                except NameError:
+                    raise NameError('%s not found in the state space'%condition.rhs)
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:431]:
```suggestion
                condition = type(condition)(condition.lhs, self.index_of[condition.rhs])
```

File: sympy/stats/stochastic_process_types.py (hunk 2 above)
Comment (czgdp1807) [reply sympy/stats/stochastic_process_types.py:436]: I think in the above and this check you are assuming that only one of LHS or RHS will be `RandomIndexedSymbol`?

Comment (naveensaigit) [reply sympy/stats/stochastic_process_types.py:436]: No, actually I'm using 2 ifs, so both of them will be executed if required. Or maybe even better would be - 
``` python
indexes=self.index_of
if isinstance(condition.lhs, Symbol) or isinstance(condition.lhs, Str):
     condition=type(condition)(indexes[condition.lhs],condition.rhs)
if isinstance(condition.rhs, Symbol) or isinstance(condition.rhs, Str):
     condition=type(condition)(condition.lhs,indexes[condition.rhs])
```
The problem with `condition=type(condition)(self.index_of[condition.lhs], self.index_of[condition.rhs])` is that `RandomIndexedSymbol` will not be part of the `indexes` returned by `index_of`

Comment (czgdp1807) [reply sympy/stats/stochastic_process_types.py:436]: For dictionaries there is something called, `get(key, default_value)` which returns `default_value` if `key` isn't there in the dictionary. Can you try using it here?

Comment (naveensaigit) [reply sympy/stats/stochastic_process_types.py:436]: `get` almost works, but fails in one place. If I were to give this - 
Without get - 
``` python
>>> T = Matrix([[0.5, 0.2, 0.3],[0.2, 0.5, 0.3],[0.2, 0.3, 0.5]])
>>> Y = DiscreteMarkovChain("Y", [1, 2, 0], T)
>>> P(Eq(Y[3], 0), Eq(Y[1], 1))
0.260000000000000
```
With get - 
``` python
>>> Y = DiscreteMarkovChain("Y", [1, 2, 0], T)
>>> P(Eq(Y[3], 0), Eq(Y[1], 1))
0.360000000000000
```

Get replaces even integers with their index. I can add an `is_integer` check to solve this. Let me know what you want to do regarding this

Comment (czgdp1807) [reply sympy/stats/stochastic_process_types.py:436]: > ```python
> >>> Y = DiscreteMarkovChain("Y", [1, 2, 0], T)
> >>> P(Eq(Y[3], 0), Eq(Y[1], 1))
> 0.360000000000000
> ```

This result is correct. The ordering matters after `index_of` was added. So, 1->0, 2->1, 0->2 so, this query is actually, `P(Eq(Y[3], 2), Eq(Y[1], 0))`. The original implementation returns `0.36` as I tried on https://live.sympy.org/

Comment (naveensaigit) [reply sympy/stats/stochastic_process_types.py:436]: Actually, one test fails -
``` python
>>> TS = MatrixSymbol('T', 3, 3)
>>> YS = DiscreteMarkovChain("Y", ['One', 'Two', 3], TS)
>>> assert P(Eq(YS[3], 3), Eq(YS[1], 1)) is S.Zero
```
This fails if I use `get` as get replaces 3 with its index 2, giving probability `T[0, 2]*T[1, 0] + T[1, 1]*T[1, 2] + T[1, 2]*T[2, 2]`.
What should I do?

Comment (czgdp1807) [reply sympy/stats/stochastic_process_types.py:436]: If noticed carefully, `ContinuousMarkovChain` isn't having `index_of` which means we cannot do the same stuff with `ContinuousMarkovChain`. Could you please allow the same features for it?

Comment (czgdp1807) [reply sympy/stats/stochastic_process_types.py:436]: A simple way could be shift `index_of` function in `MarkovProcess` or something like `FiniteStateMarkovProcess`.

Comment (czgdp1807) [reply sympy/stats/stochastic_process_types.py:436]: I see, there are several problems with the way index_of works. Can you please allow `Allow edit from maintainers`? I will be pushing some commits to your branch.

Comment (naveensaigit) [reply sympy/stats/stochastic_process_types.py:436]: Sure, I'll do that

Comment (naveensaigit) [reply sympy/stats/stochastic_process_types.py:436]: Please let me know if I can add anything else

**Comment by czgdp1807** [comment]
@naveensaigit Please go through the changes once and if possible, please check if documentation is consistent with the APIs. @Upabjojr Any thoughts? Continuous & Discrete time have now similar APIs and the common portion is in the parent class `MarkovProcess`.

**Comment by naveensaigit** [comment]
Sure, I'll do that

File: sympy/stats/stochastic_process_types.py [hunk 4]
```diff
@@ -25,6 +25,7 @@
 from sympy.stats.drv_types import Poisson, PoissonDistribution
 from sympy.stats.crv_types import Normal, NormalDistribution, Gamma, GammaDistribution
 from sympy.core.sympify import _sympify, sympify
+from sympy.core.symbol import Str
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:28]:
```suggestion
```

File: sympy/stats/tests/test_stochastic_process.py [hunk 5]
```diff
@@ -1,7 +1,7 @@
 from sympy import (S, symbols, FiniteSet, Eq, Matrix, MatrixSymbol, Float, And,
-                   ImmutableMatrix, Ne, Lt, Gt, exp, Not, Rational, Lambda, erf,
+                   ImmutableMatrix, Ne, Lt, Le, Gt, Ge, exp, Not, Rational, Lambda, erf,
```
Comment (czgdp1807) [inline sympy/stats/tests/test_stochastic_process.py:2]:
```suggestion
                   ImmutableMatrix, Ne, Lt, Gt, exp, Not, Rational, Lambda, erf,
```

File: sympy/stats/stochastic_process_types.py [hunk 6]
```diff
@@ -386,7 +433,7 @@ def _work_out_state_index(self, state_index, given_condition, trans_probs):
 
         # `not None` is `True`. So the old test fails for symbolic sizes.
         # Need to build the statement differently.
-        if FiniteSet(*[i for i in range(trans_probs.shape[0])]).is_subset(state_index) is False:
+        if FiniteSet(*[i for i in range(trans_probs.shape[0])]).is_subset(FiniteSet(*[j for j in state_index])) is False:
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:436]:
```suggestion
        cond1 = (not isinstance(state_index, Range) and
                 FiniteSet(*[i for i in range(trans_probs.shape[0])]).is_subset(FiniteSet(*[j for j in state_index])) is False)
        if cond1:
```

File: sympy/stats/stochastic_process_types.py [hunk 7]
```diff
@@ -678,21 +736,25 @@ class DiscreteMarkovChain(DiscreteTimeStochasticProcess, MarkovProcess):
...
+    and expectation statements. Here is a work-around using ``Str``:
 
-    >>> P(Eq(Y[3], Y.index_of['Rainy']), Eq(Y[1], Y.index_of['Cloudy'])).round(2)
+    >>> from sympy.core.symbol import Str
+    >>> P(Eq(Y[3], Str('Rainy')), Eq(Y[1], Str('Cloudy'))).round(2)
     0.36
 
+    >>> E(Y[3], Eq(Y[1], Str("Cloudy")))
+    0.38*Cloudy + 0.36*Rainy + 0.26*Sunny
+
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:747]:
```suggestion
```

**Comment by naveensaigit** [comment]
Looks like everything is in place. I noticed that `P(Eq(Str('Rainy', Y[1]), Eq(Y[2], Str('Cloudy'))))` doesn't work. Maybe we can interchange lhs and rhs too in `replace_with_index` in this case.

**Comment by czgdp1807** [comment]
Yeah, in every `Relationship` object keep `RandomIndexedSymbol` on the LHS if it already isn't. Say if LHS isn't already `RandomIndexedSymbol` just swap LHS and RHS. It should work everytime.

**Comment by naveensaigit** [comment]
One more small issue
``` python
>>> T = Matrix([[0.5, 0.2, 0.3],[0.2, 0.5, 0.3],[0.2, 0.3, 0.5]])
>>> sunny, cloudy, rainy = symbols('Sunny, Cloudy, Rainy')
>>> Y = DiscreteMarkovChain("Y", [sunny, cloudy, rainy], T)
>>> P(Eq(Y[3], rainy), Eq(Y[1], cloudy)).round(2)
0.36
>>> P(Eq(Y[3], rainy), Eq(Y[1], Str('Cloudy'))).round(2)
TypeError: Can only compare inequalities with Expr
```
https://github.com/sympy/sympy/blob/4683cb8950c422a52f0ed146fe1fdaa88f496d66/sympy/stats/stochastic_process_types.py#L784-L789
Adding a line like this
``` python
if isinstance(obj.number_of_states, Integer):
            for index, state in enumerate(obj.state_space):
                indices[state] = index
                indices[Str(str(state))] = index
        obj.index_of = indices
```
would do the job. Its not necessary, but might be useful for the end users. Let me know if we need this.

**Comment by czgdp1807** [comment]
> One more small issue

This was actually not about `Str` and `Symbol`. The issue was that states which were not present in the state space were moving ahead of the condition, `Ge(state, mat.shape[0]) is True` and appropriate wasn't raised.

File: sympy/stats/stochastic_process_types.py [hunk 8]
```diff
@@ -383,12 +430,16 @@ def _work_out_state_index(self, state_index, given_condition, trans_probs):
                         given_condition.atoms(RandomIndexedSymbol))
             if len(rand_var) == 1:
                 state_index = rand_var[0].pspace.set
+                print(state_index)
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:433]:
```suggestion
```

File: sympy/stats/tests/test_stochastic_process.py [hunk 9]
```diff
@@ -101,7 +102,7 @@ def test_DiscreteMarkovChain():
     raises(ValueError, lambda: E(Y[2], Eq(Y[3], 1)))
 
 
-    # extended tests for probability queries
+    # extended tests for probability queriess
```
Comment (czgdp1807) [inline sympy/stats/tests/test_stochastic_process.py:105]:
```suggestion
    # extended tests for probability queries
```

**Comment by czgdp1807** [comment]
> P(Eq(Str('Rainy', Y[1]), Eq(Y[2], Str('Cloudy'))))

Now this should work.

**Comment by czgdp1807** [comment]
If there aren't any other corner cases then we can merge this by tomorrow.

**Comment by naveensaigit** [comment]
I think all the cases are working. You can merge it! Also, I'll open a new issue to support multivariate expressions

**Comment by czgdp1807** [comment]
@Upabjojr @Smit-create Any thoughts/comments?

**Comment by naveensaigit** [comment]
Is there a way to query like this - `P(Y[3], Eq(Y[0], 1))` which gives probability of all the states at that time?

**Comment by czgdp1807** [comment]
Probably not. But what you want can be done by `T**3 * Matrix([0, 1, ..., 0])`. Given the initial distribution, find the distribution at time t.

**Comment by naveensaigit** [comment]
Yes. Will it be useful to add this too?

**Comment by asmeurer** [comment]
> BREAKING CHANGE: ContinuousMarkovChain.state_space now returns a Tuple or a Range instead of a FiniteSet.

What is the motivation behind this change? Range is a Set object, but Tuple is not (it won't work with any set operations).

**Comment by czgdp1807** [comment]
Yes. But we need to think on the API for this. For the one you are proposing probably an extra check will be needed,

```python
if isinstance(condition, RandomIndexedSymbol):
    ...
```
Or otherwise a separate function to take the initial distribution as a row Matrix and return another row matrix for the required distribution. 
Or,
density(Y[3], Eq(Y[0], 1)) as we are computing the density of Y[3] and not probability.

**Comment by naveensaigit** [comment]
> density(Y[3], Eq(Y[0], 1)) as we are computing the density of Y[3] and not probability.

Yeah, we can implement it if required

**Comment by czgdp1807** [comment]
> What is the motivation behind this change? Range is a Set object, but Tuple is not (it won't work with any set operations).

#20042 returned Tuple for `DiscreteMarkovChain` and this one keeps `ContinuousMarkovChain` consistent with that change. Though since, `index_of` stores the order information, I believe the both `DiscreteMarkovChain` and `ContinuousMarkovChain` can now return `Set` which was the case originally. I think this should be done here itself.

**Comment by czgdp1807** [comment]
> Range is a Set object, but Tuple is not (it won't work with any set operations).

Restored the type. No breaking changes for next version.

**Comment by Maelstrom6** [comment]
Thanks for implementing it for `ContinuousMarkovChain` too. I'll bump up my relevant future PRs from `DiscreteMarkovChain` to `MarkovChain` in this case.

> What is the motivation behind this change? Range is a Set object, but Tuple is not (it won't work with any set operations).

The ordering of the state space should be preserved. This is because `DiscreteMarkovChain('X', [0, 1, 2])` should be distinct from `DiscreteMarkovChain('X', [2, 1, 0])`. Some of my future methods that I'll add also require some form of ordering. This is because they are designed to return a reordered state space. But as @czgdp1807 suggests, I think I can adjust it to use `index_of`.

> Restored the type. No breaking changes for next version.

Is there a way to remove my breaking changes release notes for #20042?

**Comment by czgdp1807** [comment]
> Is there a way to remove my breaking changes release notes for #20042?

We can do it manually. Internally `state_index` is a ordered and is used instead of `state_space`. Only when state_space is called by the end user, it casts the tuple to `FiniteSet`.

**Comment by czgdp1807** [comment]
> But as @czgdp1807 suggests, I think I can adjust it to use `index_of`.

You can use, `self._state_index`. It's exactly the `Tuple`/`Range` object which you added in your previous PR.

**Comment by Maelstrom6** [comment]
> Internally `state_index` is a ordered and is used instead of `state_space`. Only when state_space is called by the end user, it casts the tuple to `FiniteSet`.

If `_state_index` returns the state space ('sunny', 'cloudy', 'rainy') instead of state indexes (0, 1, 2), why not call it `_state_space`? That way we have a private and public version of the method for its various uses.

**Comment by czgdp1807** [comment]
> If `_state_index` returns the state space ('sunny', 'cloudy', 'rainy') instead of state indexes (0, 1, 2), why not call it `_state_space`? That way we have a private and public version of the method for its various uses.

Renaming isn't a big deal. There are more interesting and challenging things to work upon to improve the APIs of stochastic processes. Let's focus on that first. Though if you find anything in-correct just report here or open an issue. 

P.S. This PR will be merged by tomorrow. Thanks.

**Comment by czgdp1807** [comment]
Merging.

**Pull request merged**